Datetime format fixes (encoding)

MiYanni opened the pull request:
Fixes: https://github.com/Azure/autorest.csharp/issues/302
Fixes: https://github.com/Azure/autorest.csharp/issues/303
Fixes: https://github.com/Azure/autorest.csharp/issues/304

This fixes the encoding for the datetime types. It does not do decoding (will do in separate PR). Unblocked a few more tests in body-complex.

MiYanni commented on src/AutoRest.CSharp.V3/Pipeline/Extensions.cs:
```diff
@@ -58,15 +66,29 @@ internal static class Extensions
+        {
+            FrameworkTypeReferenceFormat.DateTimeRFC1123 => "R",
+            FrameworkTypeReferenceFormat.DateTimeISO8601 => "yyyy-MM-ddTHH:mm:ssZ",
+            FrameworkTypeReferenceFormat.Date => "yyyy-MM-dd",
```
These formats are the format of the datetimes of the testserver tests. Do we even know if these are the right formats for all swaggers?

pakrym replied:
Nope, we don't. We'll eventually figure out if we need more.

MiYanni commented on test/AutoRest.TestServer.Tests/body-complex.cs:
```diff
@@ -274,18 +270,17 @@ public void CheckComplexPrimitiveDouble()
         {
             var result = await PrimitiveOperations.GetDateTimeRfc1123Async(ClientDiagnostics, pipeline, host);
-            Assert.AreEqual(DateTimeOffset.Parse("Mon, 01 Jan 0001 12:00:00 GMT"), result.Value.Field);
+            Assert.AreEqual(DateTimeOffset.Parse("Mon, 01 Jan 0001 00:00:00 GMT"), result.Value.Field);
```
I had the actual value wrong initially because I used the description value, not the one from the actual testserver script.

pakrym commented on src/AutoRest.CSharp.V3/ClientModels/FrameworkTypeReference.cs:
```diff
@@ -9,11 +9,13 @@ internal class FrameworkTypeReference: ClientTypeReference
+        public FrameworkTypeReferenceFormat Format { get; }
 
-        public FrameworkTypeReference(Type type, bool isNullable = false)
+        public FrameworkTypeReference(Type type, bool isNullable = false, FrameworkTypeReferenceFormat format = FrameworkTypeReferenceFormat.Default)
```
Let's put this onto the ClientObjectProperty. I'd like to keep type and serialization information separate.

MiYanni replied:
Done. Added RequestBody so that both ClientObjectProperty and RequestBody have the right information to do formatting during serialization generation.

pakrym commented on src/AutoRest.CSharp.V3/ClientModels/FrameworkTypeReferenceFormat.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace AutoRest.CSharp.V3.ClientModels
+{
+    internal enum FrameworkTypeReferenceFormat
```
Can we share the SerializationFormat enum?

MiYanni replied:
I had mentioned this before... when it was just a header format. I'll reuse it. I wrote these changes before that change.

MiYanni replied:
Fixed.

pakrym commented on test/TestServerProjects/body-complex/Generated/Models/DateWrapper.Serialization.cs:
```diff
@@ -14,12 +14,12 @@ internal void Serialize(Utf8JsonWriter writer)
             {
                 writer.WritePropertyName("field");
-                writer.WriteStringValue(Field.ToString());
+                writer.WriteStringValue(Field.Value.ToString("yyyy-MM-dd"));
```
Let's add an extension method for this. Same way we did with query/path/headers.

MiYanni replied:
I added an extension method for Utf8JsonWriter called `WriteDateTimeOffsetValue`. This allows you to pass the format specifier to it. Then, that method calls the ToString in TypeFormatters, which is the same logic used for query/path/headers.

pakrym commented on src/AutoRest.CSharp.V3/Pipeline/Extensions.cs:
```diff
@@ -28,6 +28,14 @@ internal static class Extensions
             _ => null
         };
 
+        public static FrameworkTypeReferenceFormat ToFrameworkTypeReferenceFormat(this Schema schema) => schema switch
```
This is very similar to GetSerializationFormat method

MiYanni replied:
Fixed. Used only GetSerializationFormat.

pakrym commented on src/AutoRest.CSharp.V3/Pipeline/Extensions.cs:
```diff
@@ -58,15 +66,29 @@ internal static class Extensions
+        private static Func<string, bool, FrameworkTypeReferenceFormat, string?> StringSerializer(bool includeToString = false) =>
+            (vn, nu, f) => $"writer.WriteStringValue({vn}{(includeToString ? ".ToString()" : string.Empty)});";
+
+        private static string? ToFormatSpecifier(this FrameworkTypeReferenceFormat format) => format switch
```
Let's unify this with the WriteSerializationFormat

MiYanni replied:
So, the actual formats in WriteSerializationFormat are different than this. Do header/query params have a different format than the body properties?

pakrym replied:
They are not different, they are all endup calling into https://github.com/Azure/autorest.csharp/blob/feature/v3/src/assets/Azure.Core.Shared/TypeFormatters.cs#L17 with formats exactly like here.

MiYanni replied:
I was confused that there are specifiers in WriteSerializationFormat that are specifiers for the extensions, not specifiers for the ToString on DateTimeOffset. My bad.

MiYanni replied:
So, this is a little complicated right now. Basically, I put the format specifier code (switch statement) from WriteSerializationFormat here. Why? Because, right now, the code to do this serialization logic is in Extensions. And, Extensions doesn't use the actual CodeWriter; it just returns strings. When we move all the Json-specific serialization logic out of Extensions, we can use the same WriteSerializationFormat method. For now, I did the best I can without that refactoring.

MiYanni commented on src/AutoRest.CSharp.V3/Pipeline/Extensions.cs:
```diff
@@ -58,15 +59,33 @@ internal static class Extensions
+            var valueText = $"{vn}{(nu ? ".Value" : string.Empty)}";
+            var formatText = formatSpecifier != null ? $", \"{formatSpecifier}\"" : string.Empty;
+            //TODO: Hack to call Azure.Core functionality without having the context of the namespaces specified to the file this is being written to.
+            return $"Azure.Core.Utf8JsonWriterExtensions.WriteDateTimeOffsetValue(writer, {valueText}{formatText});";
```
This is only temporary, because all this Json-specific serialization code needs to be moved into a CodeWriter. When it is, we'll have the context of namespaces/types, and without that right now, I don't know if the files have `using Azure.Core;`. For example, Model files generally don't use Azure.Core. The operation files use Azure.Core, which is why path/query/header calls to the extension methods even work right now.

pakrym replied:
What about adding a UseNamespace() method to the writer? I think it would endup being a one-liner.

pakrym replied:
I see you don't have the writer here. OK this is fine then

MiYanni commented:
@pakrym I think I've addressed all the comments. Also, there are 3 tests failing... which is odd because those tests are ignored. Any ideas?
```
Test Run Failed.
  ! UrlStringNullAsync [< 1ms]
  ! UrlStringUnicodeAsync [< 1ms]
  ! UrlUrlDateNullAsync [< 1ms]
```

https://github.com/Azure/autorest.csharp/blob/1e868271db1c44a04657a4bfc481a10fdf373f60/test/AutoRest.TestServer.Tests/url.cs#L30-L32
https://github.com/Azure/autorest.csharp/blob/1e868271db1c44a04657a4bfc481a10fdf373f60/test/AutoRest.TestServer.Tests/url.cs#L34-L36
https://github.com/Azure/autorest.csharp/blob/1e868271db1c44a04657a4bfc481a10fdf373f60/test/AutoRest.TestServer.Tests/url.cs#L56-L58

pakrym commented:
> @pakrym I think I've addressed all the comments. Also, there are 3 tests failing... which is odd because those tests are ignored. Any ideas?

Those are not the failed tests. Scroll up and you'll see the one that actually fails.

pakrym commented on src/AutoRest.CSharp.V3/ClientModels/RequestBody.cs:
```diff
@@ -0,0 +1,23 @@
+        {
+            Value = value;
+            Format = format;
+            Name = Value.IsConstant ? Value.Constant.ToValueString() : Value.Parameter.Name;
```
`ToValueString()` is an evil method and would give a wrong result in a lot of cases. Something like `WriteConstant` should be used.

But in general we shouldn't be doing codegen on this level so I would drop name property and move this logic to the code generator itself.

MiYanni replied:
There is only 1 place right now where RequestBody was being written. And, it did both the name and type calculation right before it was used. So, I thought moving them here made sense. But, I can move name back there. I'd love to use `WriteConstant`. However, that will have to wait until we remove all the serialization writing out of Extensions.cs. That's where the name value feeds into right now. :-/

MiYanni replied:
Fixed.

pakrym commented on src/AutoRest.CSharp.V3/ClientModels/RequestBody.cs:
```diff
@@ -0,0 +1,23 @@
+            Value = value;
+            Format = format;
+            Name = Value.IsConstant ? Value.Constant.ToValueString() : Value.Parameter.Name;
+            Type = Value.IsConstant ? Value.Constant.Type : Value.Parameter.Type;
```
Type property is nice but let's move it to `ConstantOrParameter` I think there are a lot of callsites that would benefit from it.

MiYanni replied:
Done.

pakrym commented on src/assets/Azure.Core.Shared/Utf8JsonWriterExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    internal static class Utf8JsonWriterExtensions
+    {
+        public static void WriteDateTimeOffsetValue(this Utf8JsonWriter writer, DateTimeOffset value, string format)
```
Why not make this method an overload of WriteStringValue?

MiYanni replied:
Great idea. Done.
